labrad/wrappers.py

```diff
@@ -393,76 +393,9 @@ def _send(self, *args, **kw):
-    else:
-        def connect():
-            return protocol.factory.connectTCP(host, port, timeout=C.TIMEOUT)
-
-        @inlineCallbacks
-        def start_tls(p, cert_string=None):
-            try:
-                resp = yield p.sendRequest(C.MANAGER_ID, [(1L, ('STARTTLS', host))])
-            except Exception, e:
-                raise Exception(
-                    'Failed sending STARTTLS command to server. You should '
-                    'update the manager and configure it to support encryption '
-                    'or else disable encryption for clients. See '
-                    'https://github.com/labrad/pylabrad/blob/master/CONFIG.md')
-            cert = resp[0][1]
-            p.transport.startTLS(crypto.tls_options(host, cert_string=cert_string))
-            returnValue(cert)
-
-        def ping(p):
-            return p.sendRequest(C.MANAGER_ID, [(2L, 'PING')])
-
-        p = yield connect()
-        is_local_connection = util.is_local_connection(p.transport)
-        if ((tls == 'starttls-force') or
-            (tls == 'starttls' and not is_local_connection)):
-            try:
-                cert = yield start_tls(p)
-            except Exception:
-                # TODO: remove this retry. This is a temporary fix to support
-                # compatibility until TLS is fully deployed.
-                print ('STARTTLS failed; will retry without encryption in case '
-                       'we are connecting to a legacy manager.')
-                p = yield getConnection(host, port, name, password, tls='off')
-                print 'Connected without encryption.'
-                returnValue(p)
-            try:
-                yield ping(p)
-            except Exception:
-                print 'STARTTLS failed due to untrusted server certificate:'
-                print 'SHA1 Fingerprint={}'.format(crypto.fingerprint(cert))
-                print
-                while True:
-                    ans = raw_input(
-                            "Accept server certificate for host '{}'? "
-                            "(accept just this [O]nce; [S]ave and always "
-                            "accept this cert; [R]eject) ".format(host))
-                    if ans.lower() in ['o', 's', 'r']:
-                        break
-                    else:
-                        print 'Invalid input:', ans
-                if ans.lower() == 'r':
-                    raise
-                p = yield connect()
-                yield start_tls(p, cert)
-                yield ping(p)
-                if ans.lower() == 's':
-                    # save now that we know TLS succeeded,
-                    # including hostname verification.
-                    crypto.save_cert(host, cert)
-    if password is None:
-        password = getPassword()
-    yield p.loginClient(password, name)
+    p = yield protocol.connect(host, port, tls)
+    yield p.authenticate(password)
```

Comment:
I want to allow for different means of authentication, some of which may require different kinds of credentials or even multiple steps (e.g. OpenID so you can use google account credentials), so I made this a separate step from just establishing a connection.